contracts/utils/Swap.sol

```diff
@@ -0,0 +1,454 @@
+    function nonceUsed(address signer, uint256 nonce)
+        public
+        view
+        override
+        returns (bool)
+    {
+        uint256 groupKey = nonce / 256;
+        uint256 indexInGroup = nonce % 256;
+        return (_nonceGroups[signer][groupKey] >> indexInGroup) & 1 == 1;
+    }
+
+    /************************************************
+     *  INTERNAL FUNCTIONS
+     ***********************************************/
+
+    /**
+     * @notice Swap Atomic ERC20 Swap
+     * @param offer Offer struct containing offer details
+     * @param bid Bid struct containing bid details
+     */
+    function _swap(Offer storage offer, Bid calldata bid) internal {
+        require(DOMAIN_CHAIN_ID == getChainId(), "CHAIN_ID_CHANGED");
+
+        address signatory = _getSignatory(bid);
+        require(signatory != address(0), "SIGNATURE_INVALID");
+        require(signatory == bid.signerWallet, "SIGNATURE_MISMATCHED");
+        require(_markNonceAsUsed(signatory, bid.nonce), "NONCE_ALREADY_USED");
+        require(offer.availableSize > 0, "ZERO_AVAILABLE_SIZE");
+        require(bid.buyAmount >= offer.minBidSize, "BID_TOO_SMALL");
+        require(bid.buyAmount <= offer.totalSize, "BID_EXCEED_TOTAL");
+
+        // Ensure min. price is met
+        uint256 bidPrice =
+            uint256(bid.sellAmount)
+                .mul(uint256(10)**offer.offeredTokenDecimals)
+                .div(bid.buyAmount);
+        require(bidPrice >= offer.minPrice, "PRICE_TOO_LOW");
+
+        // Check for partial fill
+        uint256 buyAmount = bid.buyAmount;
+        uint256 sellAmount = bid.sellAmount;
+        if (bid.buyAmount > offer.availableSize) {
+            buyAmount = offer.availableSize;
+            sellAmount = buyAmount.mul(bidPrice).div(
+                uint256(10)**offer.offeredTokenDecimals
+            );
+        }
+
+        // Transfer token from sender to signer
+        IERC20(offer.offeredToken).safeTransferFrom(
+            offer.seller,
+            bid.signerWallet,
+            buyAmount
+        );
+
+        // Transfer to referrer if any
+        uint256 feeAmount;
+        if (bid.referrer != address(0)) {
+            uint256 feePercent = referralFees[bid.referrer];
+
+            if (feePercent > 0) {
+                feeAmount = sellAmount.mul(feePercent).div(MAX_PERCENTAGE);
+
+                IERC20(offer.biddingToken).safeTransferFrom(
+                    bid.signerWallet,
+                    bid.referrer,
+                    feeAmount
+                );
+            }
+        }
+
+        // Transfer token from signer to recipient
+        IERC20(offer.biddingToken).safeTransferFrom(
+            bid.signerWallet,
+            offer.seller,
+            sellAmount.sub(feeAmount)
+        );
+
+        offer.availableSize -= uint128(buyAmount);
+        offer.totalSales += uint128(sellAmount);
+
+        if (bidPrice > offer.highestPrice) {
+            offer.highestPrice = uint128(bidPrice);
+        }
```

Comment:
All the logic for updating the state needs to happen before the ERC20 interactions, this is to avoid reentrancy. See the [check-effects-interactions pattern](https://fravoll.github.io/solidity-patterns/checks_effects_interactions.html).